packages/gatsby/src/utils/graphql-typegen/ts-codegen.ts

```diff
@@ -0,0 +1,136 @@
+import * as fs from "fs-extra"
+import { join } from "path"
+import { slash } from "gatsby-core-utils"
+import { codegen } from "@graphql-codegen/core"
+import { Kind } from "graphql"
+import type { Types } from "@graphql-codegen/plugin-helpers"
+import type { TypeScriptPluginConfig } from "@graphql-codegen/typescript/config"
+import type { TypeScriptDocumentsPluginConfig } from "@graphql-codegen/typescript-operations/config"
+import { AnyAction, Store } from "redux"
+import { CodeFileLoader } from "@graphql-tools/code-file-loader"
+import { loadDocuments } from "@graphql-tools/load"
+import { IGatsbyState, IStateProgram } from "../../redux/types"
+import { filterTargetDefinitions, stabilizeSchema } from "./utils"
+
+const OUTPUT_PATH = `src/gatsby-types.d.ts`
+const NAMESPACE = `Queries`
+
+const DEFAULT_TYPESCRIPT_CONFIG: Readonly<TypeScriptPluginConfig> = {
+  avoidOptionals: true,
+  immutableTypes: true,
+  // TODO: Better maybeValue
+  maybeValue: `T | undefined`,
+  noExport: true,
+  enumsAsTypes: true,
+  scalars: {
+    Date: `string`,
+    JSON: `any`,
+  },
+  useTypeImports: true,
+}
+
+const DEFAULT_TYPESCRIPT_OPERATIONS_CONFIG: Readonly<TypeScriptDocumentsPluginConfig> =
+  {
+    ...DEFAULT_TYPESCRIPT_CONFIG,
+    exportFragmentSpreadSubTypes: true,
+  }
+
+export async function writeTypeScriptTypes(
+  directory: IStateProgram["directory"],
+  store: Store<IGatsbyState, AnyAction>
+): Promise<void> {
+  const pluginConfig: Pick<Types.GenerateOptions, "plugins" | "pluginMap"> = {
+    pluginMap: {
+      add: require(`@graphql-codegen/add`),
+      typescript: require(`@graphql-codegen/typescript`),
+      typescriptOperations: require(`@graphql-codegen/typescript-operations`),
+    },
+    plugins: [
+      {
+        add: {
+          placement: `prepend`,
+          content: `/* eslint-disable */\n`,
+        },
+      },
+      {
+        add: {
+          placement: `prepend`,
+          content: `/* THIS FILE IS AUTOGENERATED. CHANGES WILL BE LOST ON SUBSEQUENT RUNS. */\n`,
+        },
+      },
+      {
+        add: {
+          placement: `prepend`,
+          content: `declare namespace ${NAMESPACE} {\n`,
+        },
+      },
+      {
+        typescript: DEFAULT_TYPESCRIPT_CONFIG,
+      },
+      {
+        typescriptOperations: DEFAULT_TYPESCRIPT_OPERATIONS_CONFIG,
+      },
+      {
+        add: {
+          placement: `append`,
+          content: `\n}\n`,
+        },
+      },
+    ],
+  }
+
+  const { schema, definitions } = store.getState()
+  const filename = slash(join(directory, OUTPUT_PATH))
+
+  let gatsbyNodeDocuments: Array<Types.DocumentFile> = []
+  try {
+    gatsbyNodeDocuments = await loadDocuments(
+      [`./gatsby-node.ts`, `./plugins/**/gatsby-node.ts`],
+      {
+        loaders: [new CodeFileLoader()],
+      }
+    )
+  } catch (e) {
+    // These files might not exist, so just skip this
+  }
```

Comment:
What does this exactly do? Look from graphql tags? If so should we also allow js files?

Reply:
It looks for the `graphql()` function inside gatsby-node.ts so only TS files makes sense (as you'd only use the type there)

Reply:
I might not understand exactly what this does but what if I have gatsby-node.js files but my pages are in ts, don't I want these types to be available too?

Reply:
So the mapping is as following:

Page with Query <=> Types for this exact query

So when you have your pages in TS you'll get exactly the type for the query of that page.

I don't see a use case why you might want to have the type for the query inside `gatsby-node.js` inside a page :) The types for the query inside `gatsby-node` are only valuable when you have `gatsby-node.ts`.

Does that make more sense?